### ISSUE TITLE
(global-visual-line-mode t) .spacemacs line jumping bug

### ISSUE THREAD

**Issue by TheGreenHabi** [syl20bnr/spacemacs#6177]
Expected behavior:
By enabling `global-visual-line-mode` from .spacemacs file I expect to see lines without "long line marks" and to jump lines with `j` and `k` as if long lines are real lines. In other words, the same as if I enable `global-visual-line-mode` or `visual-line-mode` manually.

Observed behavior:
Lines are shown as real lines without "long line marks" but jumping still works as if it's one long line and `visual-line-mode` is not enabled. Only after I enable this mode manually again it works as expected.

My OS: Windows 10,
spacemacs dev (All current commits of May 31, 2016 ) and all packages up to date

**Comment by izahn** [comment]
Perhaps a dumb question, but why don't we just bind `j` to `evil-next-visual-line` and `k` to `evil-previous-visual-line`? As far as I can tell these work fine even when `visual-line-mode` is not on.

**Comment by sampablokuper** [comment]
This is related to #681.

**Comment by kdelwat** [comment]
@TheGreenHabi Can you confirm that this is still an issue for you? I'm unable to reproduce on the master branch.

#### System Info :computer:
- OS: gnu/linux
- Emacs: 25.1.1
- Spacemacs: 0.200.1
- Spacemacs branch: master (rev. 3706a42)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
((auto-completion :variables auto-completion-enable-snippets-in-popup t auto-completion-enable-help-tooltip t)
 (clojure :variables clojure-enable-fancify-symbols t cider-font-lock-dynamically
          '(macro core deprecated function)
          cljr-suppress-middleware-warnings t)
 emacs-lisp git html javascript markdown org python
 (shell :variables shell-default-height 30 shell-default-position 'bottom)
 (spell-checking :variables spell-checking-enable-by-default nil)
 syntax-checking themes-megapack
 (version-control :variables version-control-diff-tool 'diff-hl version-control-global-margin t))
```

**Comment by TheGreenHabi** [comment]
@kdelwat Unfortunately, the issue is still present. I'm using latest dev as of today and all packages are updated. It's not OS specific, because I observe the same behavior in Ubuntu.

**Comment by mentalisttraceur** [comment]
I have the same issue, latest master branch with emacs 25.1.1 on Debian Sid.

Fairly new to this community, so I don't exactly know what information you'd all like as far as providing minimal reproduction steps, but I don't mind taking the time to re-run through a from-scratch spacemacs install and see if I can reproduce it with minimal changes. I also tried running `(spacemacs/toggle-visual-line-navigation-on)` inside my `.spacemac`'s `user-config` instead of just setting `global-visual-line-mode` and it didn't work either (running that function interactively once spacemacs has started up works as expected, though).

For those who just want a workaround In the meantime, I've just taken the key-remapping lines from [the toggle function](https://github.com/syl20bnr/spacemacs/blob/b7e51d70aa3fb81df2da6dc16d9652a002ba5e6b/layers/+distributions/spacemacs-base/keybindings.el#L256) and added them to my user-config right after setting global-visual-line-mode and it seems to work.

    (global-visual-line-mode t)
    (evil-define-minor-mode-key 'motion 'visual-line-mode "j" 'evil-next-visual-line)
    (evil-define-minor-mode-key 'motion 'visual-line-mode "k" 'evil-previous-visual-line)

**Comment by cjpoor** [comment]
Same problem here
spacemacs 0.200.7@24.5.1
emacs Version 24.5 (9.0)
Mac OS X 10.10.5

**Comment by TheBaronHimself** [comment]
Same issue here. 

Spacemacs  0.200.9
Emacs 25.2.1
macOS 10.12.4

**Comment by fredRos** [comment]
> Perhaps a dumb question, but why don't we just always bind j to evil-next-visual-line and k to evil-previous-visual-line? As far as I can tell these work fine even when visual-line-mode is not on.

This has an impact on other modes, for example magit. With @mentalisttraceur's workaround, moving from hunk to hunk with `j,k` in the magit status window doesn't work anymore

**Comment by mentalisttraceur** [comment]
@fredRos Thanks for noting that publicly. Since posting the workaround I've not had time to come back around to this, but I've seen a couple of corner cases where things aren't ideal with my workaround, and that's definitely one of them.

**Comment by NightMachinary** [comment]
I also have this issue on emacs Mac port.

**Comment by ctholho** [comment]
I also still have this issue.

**Comment by duianto** [comment]
The following PRs (possibly modified) changes are now on the develop branch:
Add toggle visual-line-navigation-globally #11417
It can be toggled with the key binding `SPC t C-S-l`.

**Comment by mikemc** [comment]
I am running on the current develop branch and have these lines in my `dotspacemacs/user-config` section,
```
  (set-fill-column 79)
  (global-visual-line-mode)
  (global-visual-fill-column-mode)
  (spacemacs/toggle-visual-line-navigation-globally-on)
```
However, visual line navigation doesn't work until I run `SPC t C-S-l` twice after startup. The first time, I get the message that "visual-line-navigation-globally" is disabled, and visual-line-mode also appears to stop working. After doing the toggle the second time it says its enabled; and then things finally work. So it seems like spacemacs is starting with the toggle state set to enabled, but not giving the correct behavior until it is toggled off and on again.

**Comment by mentalisttraceur** [comment]
So is the "official" Spacemacs-y solution to this issue just "use `(spacemacs/toggle-visual-line-navigation-globally-on)` instead of `(global-visual-line-mode)` in your config"?

If so, then I think

1. this should be closed with that as the official solution,

2. if this is not documented anywhere, then there should probably be a new issue opened along the lines of "document Spacemacs-y alternatives to standard visual line mode functions/configuration", and

3. if there are deficits in the Spacemacs-y alternatives, like the above comment that it interferes with visual fill column mode somehow, then separate issues should be opened for those.

**Comment by duianto** [comment]
>Edit: Ok, I noticed that if I comment out the line with (global-visual-line-mode) then visual line mode and visual line navigation work, but visual-fill-column doesn't. So I'm likely misunderstanding something basic about how spacemacs/emacs configuration is working here.

It doesn't seem to work to set the fill column in the user config with:
```elisp
(set-fill-column 79)
```
Probably because the default value is set to `80` here:
https://github.com/syl20bnr/spacemacs/blob/c0525bed816ad507b28fdc21733153ad10d0ca08/layers/%2Bspacemacs/spacemacs-defaults/config.el#L112

But it works to change it with:
```
  (setq-default fill-column 79)
```

---

mentalisttraceur, thanks for suggesting next actions for this thread.
This got lost in the saved notifications,
but I stumbled over this now and had time to test it.

>If so, then I think
>this should be closed with that as the official solution,

The bot beat us to it.

>So is the "official" Spacemacs-y solution to this issue just "use (spacemacs/toggle-visual-line-navigation-globally-on) instead of (global-visual-line-mode) in your config"?

and:
>if this is not documented anywhere, then there should probably be a new issue opened along the lines of "document Spacemacs-y alternatives to standard visual line mode functions/configuration", and

There is an entry in the FAQ:
https://github.com/syl20bnr/spacemacs/blob/develop/doc/FAQ.org#enable-navigation-by-visual-lines
>Enable navigation by visual lines?

It mentions adding:
```elisp
(spacemacs/toggle-visual-line-navigation-globally-on)
```
to the user-config.


>if there are deficits in the Spacemacs-y alternatives, like the above comment that it interferes with visual fill column mode somehow, then separate issues should be opened for those.

The suggestion above of using `setq-default` to set the fill-column seems to be the solution to mikemc's issue.
